pipeline/transformer/transformer_calm/src/main/scala/uk/ac/wellcome/platform/transformer/calm/Worker.scala

```diff
@@ -1,38 +1,34 @@
 package uk.ac.wellcome.platform.transformer.calm
 
-import uk.ac.wellcome.messaging.sns.NotificationMessage
-import uk.ac.wellcome.messaging.sqs.SQSStream
-import uk.ac.wellcome.json.JsonUtil._
-import uk.ac.wellcome.platform.transformer.calm.service.{
-  CalmTransformerWorkerServiceState,
-  MessageReadError
-}
-import uk.ac.wellcome.storage.Version
-
-trait Worker[Store, StoreKey, SourceData, SinkData] extends Runnable {
-  type Result[T] = Either[Throwable, T]
+import akka.{Done, NotUsed}
+import akka.stream.scaladsl.{Flow, Sink, Source}
 
-  def runStream(stream: SQSStream[NotificationMessage]) =
-    stream.foreach(this.getClass.getSimpleName, processMessage)
+import scala.concurrent.Future
 
-  def processMessage(message: NotificationMessage) =
-    for {
-      key <- readMessage(message)
-      storeRecord <- get(key)
-      sinkData <- doWork(storeRecord)
-      sent <- done(sinkData)
-    } yield sent
+trait Worker[Message, MessageData, SinkData] {
+  private type Result[T] = Either[Throwable, T]
+  def errorSink: Sink[Result[_], Future[Done]] = Sink.ignore
 
-  def get(storeKey: StoreKey): Result[SourceData]
+  def processMessage(source: Source[Message, NotUsed]) =
+    source.via(decodeMessageFlow).via(workFlow).via(doneFlow)
 
-  private def readMessage(message: NotificationMessage): Result[StoreKey] =
-    fromJson[StoreKey](message.body).toEither match {
-      case Left(err) =>
-        Left(new Exception(s"Worker: Cannot read message, ${err}"))
-      case _ => _
+  private def divertEither[I, O](
+    flow: Flow[I, Result[O], NotUsed],
+    to: Sink[Result[_], Future[Done]]): Flow[I, O, NotUsed] =
+    flow.divertTo(Sink.ignore, _.isLeft).collect {
+      case Right(out) => out
     }
```

Comment:
Nice. Similar to `catchErrors` [here](https://github.com/wellcomecollection/catalogue/blob/master/common/big_messaging/src/main/scala/uk/ac/wellcome/bigmessaging/FlowOps.scala#L42) fyi

Reply:
Wait a min, I don't think you need the `divertTo` here, the `collect` would filter out any `Left(_)` anyway

Reply:
It filters them out, but I want a place for them to be to use later (which I am admittedly not using, but will).

Then I could run an error flow on the error sink.

Reply:
:+1: yup if the sink wasnt just ignore then makes sense